### PR TITLE
Add Gemma2 9B Config, Sliding Window Attention and attention logits soft cap.

### PR DESCRIPTION
**Pull request by gagika**
- Adding Gemma2 9B config
- Sliding window attention support
- Attention logits soft cap

### CODE REVIEW

File: MaxText/scratch_code/gemma
Comment (khatwanimohit) [inline MaxText/scratch_code/gemma:1]: Are you using this gemma module somewhere ?

File: end_to_end/tpu/gemma/9b/test_gemma.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# export UNSCANNED_CKPT_PATH=gs://runner-maxtext-logs/2024-07-17-23-48//unscanned_chkpt/checkpoints/0/items
+
+JAX_PLATFORMS=cpu python MaxText/generate_param_only_checkpoint.py MaxText/configs/base.yml async_checkpointing=false base_output_directory=${BASE_OUTPUT_PATH} load_parameters_path=${CONVERTED_CHECKPOINT} run_name=${RUN_NAME} model_name=gemma-9b force_unroll=true
+echo "Written MaxText compatible unscanned checkpoint to ${BASE_OUTPUT_PATH}/${MODEL_VARIATION}/${RUN_NAME}/checkpoints/0/items"
+
+# We run decoding on the `UNSCANNED_CKPT_PATH` for efficient decoding on the unscanned version of the checkpoint. Note that this checkpoint only has parameters and no optimizer state. 
+# So, we use it by specifying`load_parameters_path=${CONVERTED_CHECKPOINT}`
+
+
+python MaxText/decode.py MaxText/configs/base.yml tokenizer_path=assets/tokenizer.gemma load_parameters_path=${UNSCANNED_CKPT_PATH} per_device_batch_size=1 run_name=runner_$(date +%Y-%m-%d-%H-%M) max_prefill_predict_length=8 max_target_length=16 dataset_type=synthetic steps=10 async_checkpointing=false scan_layers=false model_name=gemma-9b attention=dot_product prompt="I love to"
```
Comment (khatwanimohit) [inline end_to_end/tpu/gemma/9b/test_gemma.sh:53]: Can we split the test files in 1_test_gemma.sh and 2_test_gemma.sh. 
The first file can contain only commands related to chkpt conversion which can be run on CPU. 
Second file can contain all the train and decode commands runnable on v4-16 ( if the model doesn't fit on v4-8).

Reference: https://github.com/google/maxtext/tree/main/end_to_end/tpu/gemma/7b

**Review by khatwanimohit — APPROVED** [review]
Left a few comments. Thanks @gagika

File: MaxText/scratch_code/gemma
Comment (gagika) [reply MaxText/scratch_code/gemma:1]: No, removed it.

File: end_to_end/tpu/gemma/9b/test_gemma.sh (hunk 1 above)
Comment (gagika) [reply end_to_end/tpu/gemma/9b/test_gemma.sh]: Done.

**Review by ZhaoyueCheng — APPROVED** [review]
LGTM

File: MaxText/pyconfig.py [hunk 2]
```diff
@@ -64,19 +65,26 @@ def validate_kv_quant_axis(s: str, quantize_kvcache: bool) -> None:
...
     raise ValueError("kv_quant_axis can not be '' when quantize_kvcache is True")
 
+def validate_attention_kernel(s: str) -> None:
+  valid_attention_kernels = ("autoselected", "dot_product", "flash", "cudnn_flash_te")
+  if s not in valid_attention_kernels:  # currently supported attention
+    raise ValueError("Invalid attention kernel was passed. Valid options ", valid_attention_kernels)
+
 def validate_attention_type(s: str) -> None:
-  valid_attention_types = ("autoselected", "dot_product", "flash", "cudnn_flash_te")
+  valid_attention_types = (attention_type.value for attention_type in AttentionType)
```
Comment (ZhaoyueCheng) [inline MaxText/pyconfig.py:74]: nit maybe we can use strings to be similar to other styles in the repository?

Comment (gagika) [reply MaxText/pyconfig.py:74]: no strong preference here, what do @khatwanimohit and @gobbleturk think?

Comment (khatwanimohit) [reply MaxText/pyconfig.py:74]: Personally I like it this way. It looks more pythonic and we should have done this for the long list of model names as well.

File: end_to_end/tpu/gemma/9b/2_test_gemma.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
+
+# This file, combined with step 2 in the same directory, demonstrates converting a Gemma checkpoint from Kaggle and running various MaxText operations on it.
+# This step is tested nightly on an ordinary CPU VM.
```
Comment (khatwanimohit) [inline end_to_end/tpu/gemma/9b/2_test_gemma.sh:4]: Can you change this comment and pick it up from https://github.com/google/maxtext/blob/main/end_to_end/tpu/gemma/7b/2_test_gemma.sh

Comment (gagika) [reply end_to_end/tpu/gemma/9b/2_test_gemma.sh]: Done

File: MaxText/pyconfig.py [hunk 4]
```diff
@@ -139,6 +147,7 @@ def validate_model_name(s: str) -> bool:
       "llama3-70b",
       "mistral-7b",
       "mixtral-8x7b",
+      "gemma-9b",
```
Comment (nouiz) [inline MaxText/pyconfig.py:150]: gemma2-9b?

Comment (gagika) [reply MaxText/pyconfig.py:150]: Good call

**Comment by bzantium** [comment]
This looks like not supporting alternating global and sliding window attention used for Gemma-2.
I think implementation does not look very hard with the [reference code](https://github.com/huggingface/transformers/blob/453e74884fb7e2613e7b45033fbb3c1cadb638b4/src/transformers/models/gemma2/modeling_gemma2.py#L557).

File: MaxText/configs/base.yml [hunk 5]
```diff
@@ -135,6 +135,9 @@ remat_policy: 'full'
 scan_layers: True
 param_scan_axis: 1
 attention: 'autoselected' # Supported attention: autoselected, dot_product, flash, cudnn_flash_te
```
Comment (gobbleturk) [inline MaxText/configs/base.yml:137]: Can we add a comment clarifying the difference between attention and attention_type - which is that:
attention refers to the computation methodology
attention_type refers to different variants of the attention math
Please also list the options for attention_type here - global, local_sliding

File: MaxText/pyconfig.py (hunk 2 above)
Comment (gobbleturk) [reply MaxText/pyconfig.py:74]: Ya this looks better to me, ideally we should refactor the code to avoid strings where possible like done here

File: MaxText/tests/attention_test.py [hunk 6]
```diff
@@ -542,6 +544,113 @@ def _dot_product_attention_reshape_q(self, compute_axis_order):
 
       self.assertTrue(jax.numpy.allclose(attention_w_reshape_q_idx, attention_wo_reshape_q_idx, rtol=rtol, atol=atol, equal_nan=False))
 
+  def test_sliding_window_attention(self):
```
Comment (gobbleturk) [inline MaxText/tests/attention_test.py:547]: Thank you for including this test!

**Review by gobbleturk — APPROVED** [review]
Looks great to me, thanks Gagik! Just requested more documentation in base.yml for attention_type vs attention

File: MaxText/configs/base.yml (hunk 5 above)
Comment (gagika) [reply MaxText/configs/base.yml:139]: Done

**Comment by sultan2050** [comment]
I tried to add support to gemma2-2b by adding MaxText/configs/models/gemma2-2b.yml : 
```config
base_emb_dim: 2304
base_num_query_heads: 8
base_num_kv_heads:  4
base_mlp_dim: 9216
base_num_decoder_layers: 26
head_dim: 256
mlp_activations: ["gelu","linear"]
vocab_size: 256128
decoder_block: "gemma"
normalization_layer_epsilon: 1.e-06
logits_via_embedding: True
# final_logit_softcap: 30.0
attn_logits_soft_cap: 50.0
sliding_window_size: 4096
# use_post_attn_norm: True
# use_post_ffw_norm: True 
```

and adds "gemma2-2b"


to https://github.com/google/maxtext/blob/c35c2a3ff2b02c39bc91b19f75956d42e16a590a/MaxText/pyconfig.py#L152

Is these configuration right for 2b variant?

**Comment by salrowili** [comment]
I tried to add support to gemma2-2b by adding MaxText/configs/models/gemma2-2b.yml : 
```config
base_emb_dim: 2304
base_num_query_heads: 8
base_num_kv_heads:  4
base_mlp_dim: 9216
base_num_decoder_layers: 26
head_dim: 256
mlp_activations: ["gelu","linear"]
vocab_size: 256128
decoder_block: "gemma"
normalization_layer_epsilon: 1.e-06
logits_via_embedding: True
# final_logit_softcap: 30.0
attn_logits_soft_cap: 50.0
sliding_window_size: 4096
# use_post_attn_norm: True
# use_post_ffw_norm: True 
```

and adds "gemma2-2b"


to https://github.com/google/maxtext/blob/c35c2a3ff2b02c39bc91b19f75956d42e16a590a/MaxText/pyconfig.py#L152

Is these configuration right for 2b variant?
